universal-application-tool-0.0.1/app/views/admin/questions/QuestionEditView.java

```diff
@@ -0,0 +1,103 @@
+package views.admin.questions;
+
+import static j2html.TagCreator.body;
+import static j2html.TagCreator.br;
+import static j2html.TagCreator.form;
+import static j2html.TagCreator.label;
+
+import com.google.common.collect.ImmutableList;
+import com.google.inject.Inject;
+import j2html.tags.ContainerTag;
+import j2html.tags.DomContent;
+import java.util.Arrays;
+import java.util.Locale;
+import java.util.Optional;
+import play.mvc.Http.Request;
+import play.twirl.api.Content;
+import services.question.QuestionDefinition;
+import services.question.QuestionType;
+import services.question.TranslationNotFoundException;
+import views.BaseHtmlLayout;
+import views.BaseHtmlView;
+
+public final class QuestionEditView extends BaseHtmlView {
+  private final BaseHtmlLayout layout;
+
+  @Inject
+  public QuestionEditView(BaseHtmlLayout layout) {
+    this.layout = layout;
+  }
+
+  public Content render(Request request, Optional<QuestionDefinition> question) {
+    String headerText = question.isPresent() ? "Edit Question" : "New Question";
+    return layout.htmlContent(
+        body(
+            renderHeader(headerText),
+            buildQuestionForm(question.orElse(null)).with(makeCsrfTokenInputTag(request))));
+  }
+
+  private ContainerTag buildQuestionForm(QuestionDefinition definition) {
+    String buttonText = "";
+
+    ContainerTag formTag = form().withMethod("POST");
+
+    Optional<String> questionText = Optional.empty();
+    Optional<String> questionHelpText = Optional.empty();
+
+    if (definition != null) { // Editing a question.
+      buttonText = "Update";
+      formTag.withAction("/admin/questions/" + definition.getId());
+      formTag.with(
+          label("id: " + definition.getId()),
+          br(),
+          label("version: " + definition.getVersion()),
+          br(),
+          br());
+      try {
+        questionText = Optional.of(definition.getQuestionText(Locale.ENGLISH));
+      } catch (TranslationNotFoundException e) {
+        questionText = Optional.of("Missing Text");
+      }
+      try {
+        questionHelpText = Optional.of(definition.getQuestionHelpText(Locale.ENGLISH));
+      } catch (TranslationNotFoundException e) {
+        questionHelpText = Optional.of("Missing Text");
+      }
```

Comment:
Suggestion (blocking): I think we can extract a method for this to reduce duplication and also drop the `Optional`.

```java
protected String getLocaleString(Function<Locale, String> retrievalMethod, Locale locale) {
  try {
    return retrievalMethod.apply(locale);
  } catch (TranslationNotFoundException e) {
     logger.warn("Missing locale string for method: " + retrievalMethod.toString());
     return "Missing Text";
   }
   
... 
  questionHelpText = getLocaleString(definition::getQuestionHelpText, Locale.ENGLISH);
```

Reply:
Added Issue #237 as a follow up.